examples/enc-search-sample.rs

```diff
@@ -0,0 +1,421 @@
+// context file is named "example.devcon" in the current directory.
+//
+// Also note that this program creates two different groups - in order to avoid
+// errors when executing the program multiple times, it appends the current
+// time (as a time_t) when the program is started to the group IDs.
+//
+// Copyright (c) 2020  IronCore Labs, Inc.
+// =============================================================================
+
+use ironoxide::prelude::*;
+use lazy_static::lazy_static;
+use mut_static::MutStatic;
+use std::{
+    convert::TryFrom,
+    fmt,
+    fs::File,
+    path::PathBuf,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+type Result<T> = std::result::Result<T, AppErr>;
+
+// This is a sample struct that contains some data about a customer that an
+// application might track.
+#[derive(Clone)]
+struct Customer {
+    id: u32,
+    name: String,
+    name_keys: String,
+    email: String,
+    email_keys: String,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let sdk = initialize_sdk_from_file(&PathBuf::from("example.devcon")).await?;
+    GLOBAL_STATE.set(GlobalState::new())?;
+
+    // Create a group to use to protect the blind search index that will
+    // be used on the customer names, then create an index using that group
+    // start-snippet{createIndex}
+    let salt_group_id = create_group_id("indexedSearchGroup")?;
+    create_group(&sdk, &salt_group_id, "PII Search").await?;
+    let encrypted_salt = sdk.create_blind_index(&salt_group_id).await?;
+    let encrypted_salt_str = serde_json::to_string(&encrypted_salt)?;
+    save_encrypted_salt_to_app_server(encrypted_salt_str);
+    // end-snippet{createIndex}
+
+    // Initialize the new index so it can be used for indexing and search
+    // start-snippet{initIndex}
+    let encrypted_salt_str = get_encrypted_salt_from_app_server();
+    let encrypted_salt: EncryptedBlindIndexSalt = serde_json::from_str(&encrypted_salt_str)?;
+    let blind_index = encrypted_salt.initialize_search(&sdk).await?;
+    // end-snippet{initIndex}
+
+    // Create a group to use for the encryption of customer records.
+    // Note this is not the same as the group that was used to secure
+    // the blind index, but anyone that needs to be able to search for
+    // customers by name will need to belong to both groups.
+    let group_id = create_group_id("customerService")?;
+    create_group(&sdk, &group_id, "Customer Service").await?;
+
+    // Add some test customers to the "server store"
+    let cust1 = encrypt_customer(&sdk, 1, "Gumby", "", &group_id).await?;
+    save_customer(cust1, &[], &[]);
+    let cust2 = encrypt_customer(&sdk, 2, "Æ neid 北亰", "", &group_id).await?;
+    save_customer(cust2, &[], &[]);
+    let cust3 = encrypt_customer(&sdk, 3, "aeneid bei jing", "", &group_id).await?;
+    save_customer(cust3, &[], &[]);
+
+    // Allow the user to enter additional customers
+    let mut next_id = 4;
```

Comment:
It seems like breaking this out into a separate function would make this more readable. The reading of the values isn't key to the example.

Reply:
Done